public/app/features/explore/state/actions.ts

```diff
@@ -416,111 +424,210 @@ export function modifyQueries(
-      dispatch(clearQueriesAction({ exploreId }));
-      dispatch(stateSave()); // Remember to save to state and update location
-      return;
-    }
+const execQueries = (
+  dispatch: any,
+  getState: any,
+  exploreId: ExploreId,
+  appendQueries?: boolean,
+  direction?: QueryDirection
+) => {
+  dispatch(updateTime({ exploreId }));
+
+  const richHistory = getState().explore.richHistory;
+  const exploreItemState = getState().explore[exploreId];
+  const {
+    datasourceInstance,
+    queries,
+    containerWidth,
+    isLive: live,
+    range,
+    scanning,
+    queryResponse,
+    querySubscription,
+    history,
+    mode,
+    showingGraph,
+    showingTable,
+    prevShowMoreLogsTimeRanges,
+  } = exploreItemState;
+
+  const datasourceId = datasourceInstance.meta?.id;
+  // Some datasource's query builders allow per-query interval limits,
+  // but we're using the datasource interval limit for now
+  const minInterval = datasourceInstance.interval;
+  const datasourceName = exploreItemState.requestedDatasourceName;
+  const timeZone = getTimeZone(getState().user);
+
+  if (appendQueries && datasourceId !== 'loki') {
+    // currently append queries work only with Loki due to lack of universal line limit settings
+    // does not affect other types of queries
+    return;
+  }
 
-    // Some datasource's query builders allow per-query interval limits,
-    // but we're using the datasource interval limit for now
-    const minInterval = datasourceInstance.interval;
+  if (!hasNonEmptyQuery(queries)) {
+    dispatch(clearQueriesAction({ exploreId }));
+    dispatch(stateSave()); // Remember to save to state and update location
+    return;
+  }
 
-    stopQueryState(querySubscription);
+  // filtering out hidden queries
+  const nonHiddenQueries: DataQuery[] = queries.filter((query: DataQuery) => !query.hide);
 
-    const datasourceId = datasourceInstance.meta.id;
+  // maps queries in append queries mode to get only half of results
+  const updatedQueries: LokiQuery[] = appendQueries
+    ? nonHiddenQueries.map((query: LokiQuery) => {
+        const datasource = (datasourceInstance as unknown) as LokiDatasource;
+        const limit = query.maxLines ? query.maxLines : datasource.maxLines;
+        const maxLines = limit ? Math.floor(limit / 2) : 500;
+        return { ...query, maxLines };
+      })
+    : [];
+
+  const allResponseTimestamps: string[] = queryResponse.series
+    .map((elem: any) => {
+      return elem.fields[0].values.toArray();
+    })
+    .flat()
+    .sort((a: string, b: string) => a.localeCompare(b));
+
+  /**
+   * When using appendQueries we have to add 1s offset, with milisecond-level offsets it fetches the same log lines again
+   * can result in skipped log lines when linesLimit or maxLines params are set to the number
+   * smaller than the number of log lines during the given second.
+   */
+  const oldestResponseEndTimestamp = toUtc(allResponseTimestamps[0]).subtract('1', 's');
```

Comment:
This does not seem ideal. We are basically missing some logs in that one second aren't we? It should not even matter what the limit is I think. If last log is at time T and `from` is T0 than T0 <-> (T - 1s) means we are loosing all the logs in (T -1s) <-> T or am I getting something wrong here?

Reply:
@aocenas Yes, we do and I am also not so happy about that. The best value for that would be 1ns, but we also have a corner-case there, which I am not sure how to address (or is it way too minor issue?): when we want to display N log lines and there's at least N+1 log lines within that second, with any interval set below 1s it keeps re-fetching the same log lines.

Reply:
Maybe that's only an issue on DevEnv, but when I was manually testing it to fetch just a couple of log lines from a small interval within one second, it was re-fetching the same log lines all over again, even if they weren't within that specific narrow time range (maybe it's being casted down to seconds?).

Reply:
Well lets say we use the nanosecond timestamp here. We just want anything older that that nanosecond don't we? we need to dedup the logs in the same nanosecond but we should not just throw them away, I would rather show duplicates than drop logs.

Other issue is as you mention if we have so many logs in the same nanosecond that it fills the limit than yeah we would always fetch the same ones. With nanosecond precision the risk is bit less though and we may deal with it in some way, like extending the limit if we get the same lines 2 times in a row or something like that, but that may be something to fix later if it actually becomes an issue for someone. Still would consider bigger issue dropping logs silently than not loading more logs.

Reply:
Changed that, but only to millisecond-level precision instead of nanoseconds, because momentjs supports duration units only down to milliseconds.